Day 18: “Giving pages a clear shape by using headings” / Steve Barnett (#203)
matuzo/HTMHell: hell/adventcalendar/2025/18/index.md

---
title: "Giving pages a clear shape by using headings"
author: "Steve Barnett"
author_bio: "Steve Barnett is a human-centred front-end developer and user experience designer living in Aotearoa New Zealand. He helps software teams have happier customers by making more user-friendly software. That means making sites that everyone can use, regardless of their device, the network they’re on, or any disabilities they might have."
date: 2025-12-18
author_links:
  - label: "Personal website"
    url: "https://human-centred.nz/"
    link_label: "human-centred.nz"
  - label: "Where I work"
    url: "https://intopia.digital/"
    link_label: "Intopia"
intro: "<p>The three most common ways that headings go wonky, and how to fix them!</p>"
image: "advent25_18"
---

We can make our pages easier to understand by using headings to give our pages a clear shape. Our users might visually scan the page, use an extension or bookmarklet to list the headings, navigate using assistive technology like a screen reader, or ask AI for a summary of the page. High quality headings can make things better for everyone.

> In my day job as a Digital Accessibility Consultant, there are a couple of ways that I've seen things go a bit... wonky. Let's go through the three most common ways, and how to fix them.

Reviewer: Removing repetition of `ways`

```suggestion
In my day job as a Digital Accessibility Consultant, there are a couple of ways that I've seen things go a bit... wonky. Let's look at the three most common issues, and how to fix them.
```

Author: Nice! I prefer non-sight-specific verbs where I can 🙃 . `ways` becomes `issues` in https://github.com/matuzo/HTMHell/pull/203/commits/d90687d014459ebeafb4a21025298217b3d1615b .

## Text *should not* be a heading

Ah, this one's a real classic! When we have some big and bold text, for Design Reasons, we sometimes take a bit of a shortcut and mark it up as a heading. Let's say an `<h2>`, because it seems about the right size, or that’s what it says in the design file. But here's the thing: this text doesn't introduce or describe the content that follows. It just "needs" to be big for the look of it. 

> This is an issue because when things are marked up as headings that are not headings, it makes the page harder to understand. For example: users of assistive technology like screen readers hear things read as headings of section that are not headings.

Reviewer: The stuff after `For example` reads as a bit of a tautology, maybe something more concrete would be good here? Maybe the next section could be folded into this one for better flow. 

Or leaning into the previous example

```suggestion
This is an issue because when text is marked up as a heading even though it isn’t one, it makes the page harder to understand. For example, a screen reader user might hear:

> Heading level 3, Fresh and juicy!

They’d expect a new section of content about fresh and juicy things. But there isn’t one – it’s just a decorative slogan. That makes it harder to understand the real structure of the page and to navigate efficiently between actual sections.
```

Author: Re `Fresh and juicy!`: heh heh, I like it. I don't think it quite fits with the colours headings, though? How about `Make it pop!`? Added in https://github.com/matuzo/HTMHell/pull/203/commits/e693f9438d578dc52200743a70c267ebbb11a019 .

Author: Re "For example", "An example": hmm, yeah, it does sound repetitive 🫣 🫣 . I like having an example **section**, the more practical separate from the theory. How about this: just removing "For example:"? Done in https://github.com/matuzo/HTMHell/pull/203/commits/217a54efdf558668073e99954ce49e8604476741 .

Reviewer: Ha, simple! 🥇

### An example

Let's say we have a page explaining colours, RGB-style. We might have headings marked up as follows.

- `<h1>`Colours
  - `<h2>`Red
  - `<h2>`Green
    - `<h3>`Your new favourite colour!
  - `<h2>`Blue

In this case "Your new favourite colour!" is just some big text, designed to be eye-catching. It's not the start of a section of content.

### How to fix it

Stop using HTML and start using CSS. Instead of using a heading element, using a `<p>` or `<span>` or  `<div>` element and use CSS to make it big and bold.

## Text *should* be a heading

> Now let's come from the other side. We look at a design and see some bold text with some big, some bigger, some biggest. Sweet! We fling down a bunch of `<div>` elements, add some styles and we're done. It looks just like the design, chef's kiss, and so on. But here's the thing: this text looks like a heading, but doesn't have any semantics.

Reviewer: ```suggestion
Now let's look at the flip side. We look at a design and see some bold text. Some big, some bigger, some biggest. Sweet! We fling down a bunch of `<div>` elements, add some styles and we're done. It looks just like the design, chef's kiss, and so on. But here's the thing: this text looks like a heading, but doesn't have any semantics.
```

Author: Change the some, but left the flip in https://github.com/matuzo/HTMHell/pull/203/commits/0f1f3e13251e84caa26039a8263cd26e854499a3 . :)

> This is an issue because when things that are headings aren't marked up as headings, it makes the page harder to understand. For example: screen readers won't announce it as a heading with its level, making it hard to understand the content and structure when listening to all the headings.

Reviewer: I wonder if there's a different or additional example you could give here – I know the screen reader use case is the main one, but it feels a little repetitive. Are there other issues that could stem from lack of semantics? Maybe when the stylesheets fail to load and you're left with text that's meant to be a heading but doesn't visually stand out because it wasn't marked up as one.

Also, as above, I'd fold the example into the paragraph of re-word slightly. It threw me a bit that it says `For example: ...`, and then there's an actual `An example` section too. Is the example section an extra example, or more details about the main example, etc.

Author: Added a different example in https://github.com/matuzo/HTMHell/pull/203/commits/a815e675022918509e80649badd1bebaad9612dc .

This is an issue because users of assistive technology like screen readers use headings to understand how each section of the page relates to each other and the page as a whole. When the headings are wonky, the shape of the page is harder to understand.

### How to fix it

Use HTML to give the headings the correct nesting and ordering. Use CSS to make them look appropriately sized and shiny.

I like to start from the page as a whole and work my way down.

1. What’s the topic or purpose of this page? That text should be in an `<h1>` element near the top of the page.
2. What are the sections of the page? The name of each section should be in an `<h2>` element, at the start of the section.
3. What (if any) are the subsections of each section? The name of each subsection should be in an `<h3>` element, at the start of the subsection.
4. What (if any) are the sub-subsections of each subsection? The name of each sub-subsection should be in an `<h4>` element, at the start of the sub-subsection.

> 5. And so on, down to an `<h6>` element. Although if you've reached an `<h6>` element, it might be worth reviewing the content and seeing if there's Too Much Stuff there!

Reviewer: ```suggestion

...and so on, down to an `<h6>` element. Although if you've reached an `<h6>` element, it might be worth reviewing the content and seeing if there's Too Much Stuff there!
```

Author: Had a go at this one in https://github.com/matuzo/HTMHell/pull/203/commits/a9996537209794791ec537a08cb17e77d449d238 .

> The list of headings should read a bit like a table of contents for the page.

Reviewer: Great tip!

## Other weird heading things

There are other aspects of wonkiness that may occur. Keep a watch for these too!

> - **Heading text that doesn't make sense.** The heading text should describe and introduce the content that follows it.

Reviewer: This one is interesting and perhaps also deserving of a brief example?

Author: I'm trying to avoid examples here to keep it short and sharp (but perhaps this is not a great idea. I've added "Ask your friendly Content writer for help!" in https://github.com/matuzo/HTMHell/pull/203/commits/da8c2ec2f8db46977e91fd221045d4f8389b39e1 as a nudge towards "it's a words problem, not a markup problem". How's that? :)

Reviewer: Lovely!

- **No headings at all on the page.** There's probably some text that should be a heading.

> - **`<h1>` shenanigans: no `<h1>` element, or multiple `<h1>` elements.** Just one `<h1>` element, please! It should describe the topic or purpose of page.

Reviewer: Are there nuances here where multiple `h1`s are allowed? I'm not an expert so genuinely asking, but I see a lot of debate around single vs multiple `h1`, for example if other ones are nested inside `<article>` or `<section>` or something. Ignore if this doesn't make any sense, just wondering if there's interesting exceptions to mention.

Author: I have been a bit cheeky here and combined two issues, which perhaps makes things less clear 😬 .
Teeeeeechnically it's okay to have multiple `h1`s. The nesting doesn't materially impact the structure or interaction with the page (although it feels like it should, right?).

I think tried to err on the side of being a bit definitive here, rather than the more truthful "it depends", because we see these issues in audits, if that makes sense? Guide people in the right direction.

Reviewer: Agree, this feels like a sufficiently beginner-friendly article to not require going into nuance just yet!

> - **Skipped heading levels**, for example: jumping from an `<h2>` element to an `<h4>` element. Keep the nesting and order correct: `<h3>` elements for subsections of a section with an `<h2>` heading.

Reviewer: This feels like re-stating your third section about headings reflecting content structure, is it sufficiently different?

Author: These are subtly different from The Big Three, and things we see in accessiblity audits a bunch. I've had a go at rewording and reordering in https://github.com/matuzo/HTMHell/pull/203/commits/809fbde889666c59fa2658a9317734e0b763618d . How's that? 😄

## Accessibility nerd corner

The big three issues we started with all fall under [Web Content Accessibility Guidelines Success Criteria 1.3.1 Info and Relationships (A)](https://www.w3.org/TR/WCAG22/#info-and-relationships): "Information, structure, and relationships conveyed through presentation can be programmatically determined or are available in text."

- "Text should not be a heading" and "Text should be a heading" are "If it looks like a thing, it must be the thing in code too"

> - "Headings do not reflect the content structure" are "If it looks nested and order, it must be nested and ordered in code too"

Reviewer: I personally find these sentences with the two quoted bits a bit confusing. Neither the first or last quote seem to refer to 1.3.1?

Also, is it a little bit contradictory to what the article is saying earlier about decoupling design from semantics? Big bold text that is big and bold purely in a presentational way is valid, if it's not marked up as a heading – so it's not necessarily true that it needs to look the same way it's marked up in code?

Author: Yeah, they are confusing, good catch! I've had a go at a more clear version in https://github.com/matuzo/HTMHell/pull/203/commits/c67a162da1e57fd678cca168715bc2ec164915a0 , doing a more direct link to the WCAG words. How's that?

Reviewer: Nice! 👌

date: 2025-12-18
author_links:
  - label: "Personal website"
    url: "https://human-centred.nz/"
    link_label: "human-centred.nz"
  - label: "Where I work"
    url: "https://intopia.digital/"
    link_label: "Intopia"
intro: "<p>The three most common ways that headings go wonky, and how to fix them!</p>"
image: "advent25_18"
---

We can make our pages easier to understand by using headings to give our pages a clear shape. Our users might visually scan the page, use an extension or bookmarklet, jump around using assistive technology like a screen reader, or ask AI for a summary of the page. High quality headings can make things better for everyone.

In my day job as a Digital Accessibility Consultant, there are a couple of ways that I've seen things go a bit... wonky. Let's go through the three most common ways, and how to fix them.

## Text *should not* be a heading

> Ah, this one's a real classic! When we have some big and bold text, for Design Reasons, we sometimes take a bit of a shortcut and mark it up as a heading. Let's say an `<h2>`, because it seems about the right size, or that’s what it says in the design file. But here's the thing: this text doesn't introduce or describe the content that follows. It just "needs" to be big for the look of it. 

Reviewer 2: An example would be great, maybe in form of an outline!?

h1: Title
h2: News
h2: Products
h2: [BAD HEADING] <-- you would have to come up with something. I'm not creative enough at the moment. :)
h2: Locations

Reviewer: How about `Fresh and juicy!` 

You could imagine a little callout or something that's visually prominent but nothing to do with the heading structure.

This is an issue because when things are marked up as headings that are not headings, it makes the page harder to understand. For example: users of assistive technology like screen readers hear things read as headings of section that are not headings.

### How to fix it

Stop using HTML and start using CSS. Instead of using a heading element, using a `<p>` or `<span>` or  `<div>` element and use CSS to make it big and bold. Noice.

## Text *should* be a heading

Now let's come from the other side. We look at a design and see some bold text with some big, some bigger, some biggest. Sweet! We fling down a bunch of `<div>` elements, add some styles and we're done. It looks just like the design, chef's kiss, and so on. But here's the thing: this text looks like a heading, but doesn't have any semantics.

> This is an issue because when things that are headings aren't marked up as headings, it makes the page harder to understand. For example: screen readers won't announce it as a heading with its level, making it hard to understand the content and structure when listening to all the headings. Excellent.

Reviewer 2: ```suggestion
This is an issue because when things that are headings aren't marked up as headings, it makes the page harder to understand. For example: screen readers won't announce it as a heading with its level, making it hard to understand the content and structure when listening to all the headings.
```

### How to fix it

> Stop using CSS and start using HTML. Instead of using a `<p>` or `<span>` or `<div>` element, use a heading element at the right level to give it semantic structure: from `<h1>` to `<h6>`.

Author: Updated in 31f16e59b737185c5e14852a701062f262e250ee .

3. What (if any) are the subsections of each section? The name of each subsection should be in an `<h3>` element, at the start of the subsection.
4. What (if any) are the sub-subsections of each subsection? The name of each sub-subsection should be in an `<h4>` element, at the start of the sub-subsection.
5. And so on, down to an `<h6>` element. Although if you've reached an `<h6>` element, it might be worth reviewing the content and seeing if there's Too Much Stuff there!

The list of headings should read a bit like a table of contents for the page.

## Other weird heading things

There are other aspects of wonkiness that may occur. Keep a watch for these too!

- **No headings at all on the page.** There's probably some text that should be a heading.
- **`<h1>` shenanigans: no `<h1>` element, or multiple `<h1>` elements.** Just one `<h1>` element, please! It should describe the topic or purpose of page.
- **Skipped heading levels**, for example: jumping from an `<h2>` element to an `<h4>` element. Keep the nesting and order correct: `<h3>` elements for subsections of a section with an `<h2>` heading.

> ## Accessibility nerd corner

Author: Added some bits in 8d1a6d222f3db944efa86f9d313262e0d283a011 .
